packages/cli/src/controller/publish-controller.ts

```diff
@@ -0,0 +1,65 @@
+// Copyright 2020-2021 OnFinality Limited authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import {loadProjectManifest} from '@subql/common';
+import fs from 'fs';
+import IPFS from 'ipfs-http-client';
+import path from 'path';
+
+// https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/FILES.md#filecontent
+type FileContent = Uint8Array | String | Iterable<Uint8Array> | Iterable<number> | AsyncIterable<Uint8Array>;
+
+// https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/FILES.md#fileobject
+type FileObject = {
+  path?: string;
+  content?: FileContent;
+  mode?: number | string;
+  mtime?: Date | number[] | {secs: number; nsecs?: number};
+};
+
+type IterableType<T> = T extends AsyncIterable<infer U> ? U : T extends Iterable<infer U> ? U : T;
+
+type AddResult = IterableType<ReturnType<IPFS.IPFSHTTPClient['addAll']>>;
+
+export async function uploadToIpfs(ipfsEndpoint: string, projectDir: string): Promise<AddResult[]> {
+  const ipfs = IPFS.create({url: ipfsEndpoint});
+
+  const files = getFiles(projectDir);
+
+  const pendingResults = ipfs.addAll(files, {
+    wrapWithDirectory: true,
+    pin: true, // Theres no guarantee the server will respect this
```

Comment:
IPFS gateways have no incentive to pin (persist) content. This could mean that the project is inaccessible on IPFS. 
There are pinning services such as [Pinata](https://www.pinata.cloud/) but they do pinning via their own api.